kernel/runtime/kobject/runtime/ExecutionContext.hh

```diff
@@ -54,14 +52,28 @@ namespace mythos {
         Msg& cs(CapPtr c) { ib.cs(c); return *this; }
         Msg& cs(CapMap& c) { ib.cs(c.cap()); return *this; }
         Msg& sched(CapPtr c) { ib.sched(c); return *this; }
-        Msg& rawStack(void* ptr) { ib.regs.rsp = uintptr_t(ptr); return *this; }
+        Msg& rawStack(uintptr_t ptr) { ib.regs.rsp = ptr; return *this; }
+        Msg& rawStack(void* ptr) { return rawStack(uintptr_t(ptr)); }
+
+        // TODO: does this really belong here?
         Msg& prepareStack(void* ptr) {
-            // \TODO assert alignment
-            auto tos = reinterpret_cast<uintptr_t*>(ptr);
-            *--tos = 0;                                     // dummy return address
-            ib.regs.rsp = uintptr_t(tos); 
+            // The compiler expect a kinda strange alignment.
+            // -> rsp % 16 must be 8
+            // We do that in clone, so we better do it here too.
+            // You can see this also in musl/src/thread/x86_64/clone.s (rsi is stack)
+            // We will use the same trick for alignment as musl libc
+            ib.regs.rsp = ((uintptr_t(ptr)) & uintptr_t(-16)) - 8; 
+            *(uintptr_t*)(ib.regs.rsp) = 0; // dummy return address
```

Comment:
rsp-- missing after this,  or --rsp ?

Reply:
No, this would mess up the stack alignment. sizeof(untptr_t) is 8, so the alignment actually reserves the space on the stack. I will add a comment.